File: day16/main.go
package main

import (
	"bufio"
	"encoding/hex"
	"fmt"
	"log"
	"os"
)

const (
	puzzleInput = "input.txt"
)

type (
	BitReader struct {
		bit    int
		offset int
		buffer []byte
	}
)

func NewBitReader(buffer []byte) *BitReader {
	return &BitReader{
		bit:    0,
		offset: 0,
		buffer: buffer,
	}
}

func (r *BitReader) readBit() (byte, bool) {
	if r.offset >= len(r.buffer) {
		return 0, false
	}
	b := r.buffer[r.offset]
	k := (b >> (7 - r.bit)) & 1
	r.bit = (r.bit + 1) % 8
	if r.bit == 0 {
		r.offset++
	}
	return k, true
}

func (r *BitReader) ReadBits(a int, b []byte) (n int) {
	for i := range b {
		if n >= a {
			return
		}
		k, ok := r.readBit()
		if !ok {
			return
		}
		b[i] = k
		n++
	}
	return
}

func bitsToByte(b []byte) byte {
	var k byte = 0
	for _, i := range b {
		k = (k << 1) + i
	}
	return k
}

func bitsToInt(b []byte) int {
	k := 0
	for _, i := range b {
		k = (k << 1) + int(i)
	}
	return k
}

func nibblesToInt(b []byte) int {
	k := 0
	for _, i := range b {
		k = (k << 4) + int(i)
	}
	return k
}

func parsePacketHeader(tokens []int) (int, int, []int, bool) {
	if len(tokens) < 2 {
		return 0, 0, tokens, false
	}
	return tokens[0], tokens[1], tokens[2:], true
}

func parseSubpackets0(l, offset int, tokens []int) ([]int, int, []int, bool) {
	origOffset := offset
	origTokens := tokens
	var vals []int
	for {
		if offset-origOffset >= l {
			break
		}
		var val int
		var ok bool
		val, offset, tokens, ok = evalPacket(offset, tokens)
		if !ok {
			return nil, origOffset, origTokens, false
		}
		vals = append(vals, val)
	}
	return vals, offset, tokens, true
}

func parseSubpackets1(l, offset int, tokens []int) ([]int, int, []int, bool) {
	origOffset := offset
	origTokens := tokens
	vals := make([]int, 0, l)
	for i := 0; i < l; i++ {
		var val int
		var ok bool
		val, offset, tokens, ok = evalPacket(offset, tokens)
		if !ok {
			return nil, origOffset, origTokens, false
		}
		vals = append(vals, val)
	}
	if len(vals) < l {
		return nil, origOffset, origTokens, false
	}
	return vals, offset, tokens, true
}

func evalPacket(offset int, tokens []int) (int, int, []int, bool) {
	origOffset := offset
	origTokens := tokens
	var id int
	var ok bool
	_, id, tokens, ok = parsePacketHeader(tokens)
	if !ok {
		return 0, origOffset, origTokens, false
	}
	if id == 4 {
		if len(tokens) < 2 {
			return 0, origOffset, origTokens, false
		}
		return tokens[0], tokens[1], tokens[2:], true
	}
	if len(tokens) < 3 {
		return 0, origOffset, origTokens, false
	}
	mode := tokens[0]
	l := tokens[1]
	offset = tokens[2]
	var vals []int
	if mode == 0 {
		var ok bool
		vals, offset, tokens, ok = parseSubpackets0(l, offset, tokens[3:])
		if !ok {
			return 0, origOffset, origTokens, false
		}
	} else {
		var ok bool
		vals, offset, tokens, ok = parseSubpackets1(l, offset, tokens[3:])
		if !ok {
			return 0, origOffset, origTokens, false
		}
	}
	switch id {
	case 0:
		{
			k := 0
			for _, i := range vals {
				k += i
			}
			return k, offset, tokens, true
		}
	case 1:
		{
			k := 1
			for _, i := range vals {
				k *= i
			}
			return k, offset, tokens, true
		}
	case 2:
		{
			k := vals[0]
			for _, i := range vals {
				if i < k {
					k = i
				}
			}
			return k, offset, tokens, true
		}
	case 3:
		{
			k := vals[0]
			for _, i := range vals {
				if i > k {
					k = i
				}
			}
			return k, offset, tokens, true
		}
	case 5:
		{
			k := 0
			if vals[0] > vals[1] {
				k = 1
			}
			return k, offset, tokens, true
		}
	case 6:
		{
			k := 0
			if vals[0] < vals[1] {
				k = 1
			}
			return k, offset, tokens, true
		}
	case 7:
		{
			k := 0
			if vals[0] == vals[1] {
				k = 1
			}
			return k, offset, tokens, true
		}
	}
	return 0, origOffset, origTokens, false
}

func main() {
	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatal(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatal(err)
		}
	}()

	var bitstream []byte
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		var err error
		bitstream, err = hex.DecodeString(scanner.Text())
		if err != nil {
			log.Fatal(err)
		}
	}

	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}

	var tokens []int
	bits := NewBitReader(bitstream)
	buf := make([]byte, 15)
	bitOffset := 0
	versionSum := 0
	for {
		n := bits.ReadBits(3, buf)
		if n != 3 {
			break
		}
		bitOffset += 3
		version := bitsToByte(buf[:n])
		n = bits.ReadBits(3, buf)
		if n != 3 {
			break
		}
		bitOffset += 3
		id := bitsToByte(buf[:n])
		tokens = append(tokens, int(version))
		tokens = append(tokens, int(id))
		versionSum += int(version)
		if id == 4 {
			var nibbles []byte
			for {
				n := bits.ReadBits(5, buf)
				if n != 5 {
					break
				}
				bitOffset += 5
				nibbles = append(nibbles, bitsToByte(buf[1:n]))
				if buf[0] == 0 {
					break
				}
			}
			tokens = append(tokens, nibblesToInt(nibbles))
		} else {
			n := bits.ReadBits(1, buf)
			if n != 1 {
				break
			}
			bitOffset += 1
			mode := buf[0]
			tokens = append(tokens, int(mode))
			if mode == 0 {
				n := bits.ReadBits(15, buf)
				if n != 15 {
					break
				}
				bitOffset += 15
				tokens = append(tokens, bitsToInt(buf[:n]))
			} else {
				n := bits.ReadBits(11, buf)
				if n != 11 {
					break
				}
				bitOffset += 11
				tokens = append(tokens, bitsToInt(buf[:n]))
			}
		}
		tokens = append(tokens, bitOffset)
	}
	fmt.Println("Part 1:", versionSum)
	val, _, _, ok := evalPacket(0, tokens)
	if !ok {
		log.Fatalln("Failed eval")
	}
	fmt.Println("Part 2:", val)
}

File: day16/src/main.rs
use std::collections::VecDeque;
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;

const PUZZLEINPUT: &str = "input.txt";

struct BitReader {
    bit: usize,
    offset: usize,
    buffer: VecDeque<u8>,
}

impl BitReader {
    fn new(buffer: VecDeque<u8>) -> Self {
        Self {
            bit: 0,
            offset: 0,
            buffer,
        }
    }

    fn read_bit(&mut self) -> (u8, bool) {
        if self.offset >= self.buffer.len() {
            return (0, false);
        }
        let b = self.buffer[self.offset];
        let k = (b >> (7 - self.bit)) & 1;
        self.bit = (self.bit + 1) % 8;
        if self.bit == 0 {
            self.offset += 1;
        }
        (k, true)
    }

    fn read_bits(&mut self, a: usize, b: &mut Vec<u8>) -> usize {
        b.clear();
        for _ in 0..a {
            let (k, ok) = self.read_bit();
            if !ok {
                return b.len();
            }
            b.push(k);
        }
        b.len()
    }
}

fn bits_to_byte(b: &[u8]) -> u8 {
    let mut k = 0;
    for i in b {
        k = (k << 1) + i;
    }
    return k;
}

fn bits_to_int(b: &[u8]) -> usize {
    let mut k = 0;
    for &i in b {
        k = (k << 1) + i as usize;
    }
    return k;
}

fn nibbles_to_int(b: &[u8]) -> usize {
    let mut k = 0;
    for &i in b {
        k = (k << 4) + i as usize;
    }
    return k;
}

fn parse_packet_header(tokens: &[usize]) -> Option<(usize, usize, &[usize])> {
    match tokens {
        [version, id, rest @ ..] => Some((*version, *id, rest)),
        _ => None,
    }
}

fn parse_packet_literal(tokens: &[usize]) -> Option<(usize, usize, &[usize])> {
    match tokens {
        [val, offset, rest @ ..] => Some((*val, *offset, rest)),
        _ => None,
    }
}

fn parse_subpacket_opts(tokens: &[usize]) -> Option<(usize, usize, usize, &[usize])> {
    match tokens {
        [mode, l, offset, rest @ ..] => Some((*mode, *l, *offset, rest)),
        _ => None,
    }
}

fn parse_subpackets_0(
    l: usize,
    offset: usize,
    tokens: &[usize],
) -> Option<(Vec<usize>, usize, &[usize])> {
    let mut vals = Vec::new();
    let mut cur_offset = offset;
    let mut cur_tokens = tokens;
    loop {
        if cur_offset - offset >= l {
            break;
        }
        if let Some((val, o, t)) = eval_packet(cur_tokens) {
            cur_offset = o;
            cur_tokens = t;
            vals.push(val);
        } else {
            return None;
        }
    }
    Some((vals, cur_offset, cur_tokens))
}

fn parse_subpackets_1(
    l: usize,
    offset: usize,
    tokens: &[usize],
) -> Option<(Vec<usize>, usize, &[usize])> {
    let mut vals = Vec::new();
    let mut cur_offset = offset;
    let mut cur_tokens = tokens;
    for _ in 0..l {
        if let Some((val, o, t)) = eval_packet(cur_tokens) {
            cur_offset = o;
            cur_tokens = t;
            vals.push(val);
        } else {
            return None;
        }
    }
    Some((vals, cur_offset, cur_tokens))
}

fn eval_packet(tokens: &[usize]) -> Option<(usize, usize, &[usize])> {
    let mut cur_offset;
    let mut cur_tokens = tokens;
    let id = match parse_packet_header(cur_tokens) {
        Some((_, id, t)) => {
            cur_tokens = t;
            id
        }
        None => return None,
    };
    if id == 4 {
        return parse_packet_literal(cur_tokens);
    }
    let (mode, l) = match parse_subpacket_opts(cur_tokens) {
        Some((mode, l, o, t)) => {
            cur_offset = o;
            cur_tokens = t;
            (mode, l)
        }
        None => return None,
    };
    let vals = match if mode == 0 {
        parse_subpackets_0(l, cur_offset, cur_tokens)
    } else {
        parse_subpackets_1(l, cur_offset, cur_tokens)
    } {
        Some((vals, o, t)) => {
            cur_offset = o;
            cur_tokens = t;
            vals
        }
        None => return None,
    };
    let k = match id {
        0 => vals.iter().sum::<usize>(),
        1 => vals.iter().product::<usize>(),
        2 => {
            if let Some(&v) = vals.iter().min() {
                v
            } else {
                return None;
            }
        }
        3 => {
            if let Some(&v) = vals.iter().max() {
                v
            } else {
                return None;
            }
        }
        5 => {
            if let [a, b] = vals[..] {
                if a > b {
                    1
                } else {
                    0
                }
            } else {
                return None;
            }
        }
        6 => {
            if let [a, b] = vals[..] {
                if a < b {
                    1
                } else {
                    0
                }
            } else {
                return None;
            }
        }
        7 => {
            if let [a, b] = vals[..] {
                if a == b {
                    1
                } else {
                    0
                }
            } else {
                return None;
            }
        }
        _ => return None,
    };
    Some((k, cur_offset, cur_tokens))
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let file = File::open(PUZZLEINPUT)?;
    let reader = BufReader::new(file);

    let bitstream = if let Some(line) = reader.lines().flat_map(|i| i).next() {
        (0..line.len())
            .flat_map(|i| u8::from_str_radix(&line[i..i + 1], 16))
            .collect::<Vec<_>>()
            .chunks(2)
            .flat_map(|i| match i {
                &[a, b] => Some((a << 4) + b),
                &[a] => Some(a << 4),
                _ => None,
            })
            .collect::<VecDeque<u8>>()
    } else {
        return Err("Invalid line".into());
    };

    let mut tokens = Vec::new();
    let mut bits = BitReader::new(bitstream);
    let mut buf = Vec::with_capacity(15);
    let mut bit_offset = 0;
    let mut version_sum = 0;
    loop {
        if bits.read_bits(3, &mut buf) != 3 {
            break;
        }
        bit_offset += 3;
        let version = bits_to_byte(&buf[..]);
        if bits.read_bits(3, &mut buf) != 3 {
            break;
        }
        bit_offset += 3;
        let id = bits_to_byte(&buf[..]);
        tokens.push(version as usize);
        tokens.push(id as usize);
        version_sum += version as usize;
        if id == 4 {
            let mut nibbles = Vec::new();
            loop {
                if bits.read_bits(5, &mut buf) != 5 {
                    break;
                }
                bit_offset += 5;
                nibbles.push(bits_to_byte(&buf[1..]));
                if buf[0] == 0 {
                    break;
                }
            }
            tokens.push(nibbles_to_int(&nibbles));
        } else {
            if bits.read_bits(1, &mut buf) != 1 {
                break;
            }
            bit_offset += 1;
            let mode = buf[0];
            tokens.push(mode as usize);
            if mode == 0 {
                if bits.read_bits(15, &mut buf) != 15 {
                    break;
                }
                bit_offset += 15;
                tokens.push(bits_to_int(&buf[..]));
            } else {
                if bits.read_bits(11, &mut buf) != 11 {
                    break;
                }
                bit_offset += 11;
                tokens.push(bits_to_int(&buf[..]));
            }
        }
        tokens.push(bit_offset);
    }
    println!("Part 1: {}", version_sum);
    if let Some((val, _, _)) = eval_packet(&tokens[..]) {
        println!("Part 2: {}", val);
    } else {
        return Err("Failed eval".into());
    }
    Ok(())
}
